lib/client/conntest/database.go

```diff
@@ -0,0 +1,407 @@
+			// if the test reaches the end, we force the test to be a success by calling
+			// 	connectionDiagnostic.SetMessage(types.DiagnosticMessageSuccess)
+			//	connectionDiagnostic.SetSuccess(true)
+			Message: types.DiagnosticMessageFailed,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	databaseServers, err := s.getDatabaseServers(ctx, req.ResourceName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(databaseServers) == 0 {
+		connDiag, err := s.appendDiagnosticTrace(ctx,
+			connectionDiagnosticID,
+			types.ConnectionDiagnosticTrace_RBAC_DATABASE,
+			"You are not authorized to access this Database. "+
+				"Ensure your role grants access by adding it to the 'db_labels' property. "+
+				"This can also happen when you don't have a Database Agent proxying the database - "+
+				"you can fix that by adding the database labels to the 'db_service.resources.labels' in 'teleport.yaml' file of the database agent.",
+			trace.NotFound("%s not found", req.ResourceName),
+		)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return connDiag, nil
+	}
+
+	databaseServer := databaseServers[0]
+	routeToDatabase := proto.RouteToDatabase{
+		ServiceName: databaseServer.GetName(),
+		Protocol:    databaseServer.GetDatabase().GetProtocol(),
+		Username:    req.DatabaseUser,
+		Database:    req.DatabaseName,
+	}
+
+	databasePinger, err := getDatabaseConnTester(routeToDatabase.Protocol)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := checkDatabaseLogin(routeToDatabase.Protocol, req.DatabaseUser, req.DatabaseName); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if _, err := s.appendDiagnosticTrace(ctx,
+		connectionDiagnosticID,
+		types.ConnectionDiagnosticTrace_RBAC_DATABASE,
+		"A Database Agent is available to proxy the connection to the Database.",
+		nil,
+	); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := s.runALPNTunnel(ctx, req, routeToDatabase, connectionDiagnosticID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ping, err := newPing(listener.Addr().String(), req.DatabaseUser, req.DatabaseName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if pingErr := databasePinger.Ping(ctx, ping); pingErr != nil {
+		connDiag, err := s.handlePingError(ctx, connectionDiagnosticID, pingErr, databasePinger)
+		return connDiag, trace.Wrap(err)
+	}
+
+	return s.handlePingSuccess(ctx, connectionDiagnosticID)
+}
+
+func (s *DatabaseConnectionTester) runALPNTunnel(ctx context.Context, req TestConnectionRequest, routeToDatabase proto.RouteToDatabase, connectionDiagnosticID string) (net.Listener, error) {
+	list, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = client.RunALPNAuthTunnel(ctx, client.ALPNAuthTunnelConfig{
```

Comment:
I'll call `defer listener.Close()` right after the call to run the ALPN Tunnel in the connection tester.
I overlooked this and assumed it would get closed by the context